rfcs/0166-nix-formatting.md

```diff
@@ -0,0 +1,1272 @@
+
+```nix
+#1 single line
+foo = "bar";
+
+#2 single line, on a new line
+very.long.foo =
+  function arg1 arg2 arg3;
+
+#3 multi line, starting on the same line
+foo = function {
+  # args
+};
+
+#4 multi line, starting on a new line
+foo =
+  function
+    arg1
+    arg2
+    arg3;
+
+#5 simple function arguments (<=2) start on the same line
+add = x: y: {
+  result = x + y;
+};
+
+#6 attribute bindings always start on a new line
+# If we add more attributes, the indentation stays the same, see #7
+split =
+  { x, ... }:
+  {
+    inc = x + 1;
+    dec = x - 1;
+  };
+
+#7 with enough attribute bindings, they will start on a new line and become multline 
+# it is common to expand input arguments in a way that does not change the number of function applications necessary
+outputs =
+  {
+    x,
+    y,
+    z,
+    ...
+  }:
+  {
+    result = x + y;
+  };
+```
+
+Notable special cases are:
+
+- Single line values that would not benefit from style #2 keep using #1, even if this makes it go above the line limit. This mostly applies to simple strings and paths.
+- Attribute set values must always be expanded. This has the consequence of always forcing nested attribute sets to be multiline (even if they would be single line otherwise because they only contain a single item), which usually is desired.
+  ```nix
+  {
+    foo.bar.baz = "qux";
+    foo' = {
+      bar.baz = "qux";
+    };
+  }
+  ```
+- "statement-like" expressions like "let", "if" and "assert" always use #4 (or #1).
+- If the value is a `with` followed by a function application or list or attribute set, try to use #3.
+  ```nix
+  buildInputs = with pkgs; [
+    some
+    dependencies
+  ];
+  ```
+
+**Alternatives**
+
+One could eliminate style #2 by having #4 always start on the first line. This would even reduce indentation in some cases. However, this may look really weird in other cases, especially when the binding is very long:
+
+```nix
+some.very.long.attr = callFunction
+  arg1
+  arg2
+  arg3;
+```
+
+#### Bindings semicolon placement
+
+The semicolon in bindings must always be placed on the same line as the expression it concludes.
+
+**Examples**
+
+```nix
+{
+  attr1 = bar;
+  attr2 = function call {
+    # stuff
+  };
+  attr3 =
+    function call
+      many
+      arguments;
+  attr4 =
+    let
+      foo = "bar";
+    in
+    some statement;
+  attr5 =
+    if foo then
+      "bar"
+    else
+      "baz";
+  attr6 =
+    let
+      foo = false;
+    in
+    if foo then "bar" else "baz";
+  attr7 = function (
+    if foo then
+      "bar"
+    else
+      "baz"
+  );
+  attr8 =
+    cond1
+    || cond2
+    ||
+      some function call
+      && cond3;
+}
+```
+
+**Alternatives**
+
+
+1. On a new line without indentation.
+  - This clearly marks a separation between attributes, however it is wasteful of space.
+  ```nix
+  attr3 =
+    function call
+      many
+      arguments
+  ;
+  attr3 =
+    let
+      foo = "bar";
+    in
+    some statements
+  ;
+  ```
```

Comment:
Just to clarify, you're suggesting that the formatter should _always_ add parenthesis for multi-line bindings? It sounds very controversial and non-standard, but I like the idea!

Reply:
Actually, thinking about it again, with a let binding with single line after `in` would probably be okay to have to terminator on that line because it statement is always the actual value of the binding (the let block just elaborates it).  
With other statements broken up into multiple lines such as ternaries, function calls or operator chainings, the order may change and the last line does not necessarily represent the true value of the binding.

I'm not sure about proposing it as the "chosen" style but I wanted to at least have it on the record as an alternative to think and reason about or perhaps take inspiration from. You can always choose to format it like this on your own and it'd still be valid under the new rules, even if the formatter doesn't automatically produce code like this.

Reply:
In my opinion bindings are easy to parse visually. We shouldn’t add visual noise to them. The semicolon should just be at the end of the last line of the binding. The semicolon is there to help the parser, not humans.

Reply:
If we waste a line for the `;` I don’t care very much, if we use parentheses or not. I will just note that editing a file with additional syntax requires slightly more typing.

Reply:
Now added this idea as an alternative: https://github.com/NixOS/rfcs/pull/166/commits/aaf1613f51abb97be149a5f429634a25277c2327